docs/Generators.md

```diff
@@ -596,3 +597,69 @@ The matrix generators will combine both output and produce:
           template: # Ignored
        
 ```
+
+## Cluster List Resource Generator
+
+The cluster list resource generates a list of Argo CD clusters. This is done using [duck-typing](https://pkg.go.dev/knative.dev/pkg/apis/duck), which does not require knowledge of the full shape of the referenced kubernetes resource. The following is an example of the new ApplicationSet generator:
+```yaml
+apiVersion: argoproj.io/v1alpha1
+kind: ApplicationSet
+metadata:
+ name: guestbook
+spec:
+ generators:
```

Comment:
Don't you think it will be more consistent to add `clusterListResource` under the existing `cluster` field? This looks more natural to me because `resource` is another source of clusters same as `selector` .

So the generator definition would look as follows:

```yaml
 generators:
 - clusterList:
     resource:
       configMapRef: my-configmap
       name: quak
```

Reply:
I see your point around the name choice.  The name choice for the generator does overlap, and does not infer the capability I was trying to incorporate.  There were some alternate names suggested from @jgwest maybe `clusterDecisionResource` is more true to what the duckType status on the targeted resource offers.

Thoughts?

Reply:
Hey @alexmt, were you suggesting that we move this into the existing Cluster generator? (you used `clusterList` in your example, which isn't an existing field, so it was a bit ambigous). 

But, presuming so, my preference is to keep them separate:
- Once concern I have is that the learning curve for this generator is necessarily steeper (duck types, CRDs, external controllers) than the existing cluster generator. I want to keep the learning curve on the cluster generator as simple as possible, because the existing Cluster generator will the first generator, and likely most popular, that users use.
    - For example, re: generator complexity and learning curve, I carefully ordered the list of generators [on the documentation page](https://argocd-applicationset.readthedocs.io/en/stable/Generators/), from simplest to most complex, to gradually introduce users to the generator functionality. (Admittedly, it's a small thing, but I think has an outsized impact on user engagement.)
- Generators thus far have been named for the source of the template parameters that they generate, rather than for their target: list provides a set of parameters from a literal list, cluster uses the argo cd cluster list as a source, git (from git, either file or directory), scmProvider (from an SCM provider like GitHub/GitLab) and so on.
    - Thus here it makes sense that this new generator would be named for it's data source, rather than it's target. It's datasouce is the fields from custom CRs (with the contents of those fields being mutable and potentially based on fancy, custom logic).
- IMHO this generator's functionality is more than just a simple selector: the purpose is to support custom resources which decide (using the implementation of another controller) where Application resources should exist in a cluster fleet. With this approach the other controller can use quite fancy custom logic, which is more powerful than a simple selector.


However, as @jnpacker mentioned, the generator name being proposed could be misleading. His proposal of `clusterDecisionResource` I think better captures that purpose of the custom resource, which is to implement complex decision-making re: where cluster resources should live (rather than a simple, immutable list of clusters that the old name might imply.)